### ISSUE TITLE
Updating Custom Fields?

### ISSUE THREAD

**Issue by walkeryan** [EncompassRest/EncompassRest#26]
Custom Fields don't seem to be sticking when trying to assign them. Perhaps its something I'm doing wrong. 

I've tried using GetLoanAsync and updating the CF in the loan.CustomFields object as well as creating a new loan using the Loan GUID and adding the CF. 

When I use Loans.UpdateLoanAsync The Standard Field (Property State) that I updated has the new data but the Custom Field Still has the Old data in it. Here is the Code I'm Using:

Loan loan = await client.Loans.GetLoanAsync(item.LoanGuid);

                            loan.Property.State = "OH";

                            foreach (CustomField cf in loan.CustomFields)
                            {
                                if (cf.FieldName == "CUST91FV")
                                {
                                    cf.StringValue = "Delete this!";
                                }
                            }

                            await client.Loans.UpdateLoanAsync(loan, true).ConfigureAwait(false);

**Comment by walkeryan** [comment]
Thanks for the fast response. You guys are doing excellent work. So If I use the
loan.CustomFields = loan.CustomFields
Are the business rules/ Calcs going to run on every custom field?

Thanks

**Comment by TylerBrinkley** [comment]
Possibly. I know for sure they'll run on any custom fields you've updated but they may or may not run for the custom fields whose values don't change.

**Comment by walkeryan** [comment]
OK thanks a lot for the help. I'll give it a shot.

**Comment by walkeryan** [comment]
Unfortunately I'm getting a bunch of field validation errors when trying to use the loan.CustomFields = loan.CustomFields for the Blank Fields. Is there any other way to Mark the Collection as being Dirty without assigning it to itself? Thanks


RestException: UpdateLoanAsync
{"summary":"Bad Request","details":"Invalid CustomField Id ='CustomField\/32', FieldName = 'CX.COMPARELP'\u000d\u000aInvalid CustomField Id ='CustomField\/73', FieldName = 'CX.VAF.BAS'\u000d\u000aInvalid CustomField Id ='CustomField\/82', FieldName = 'CX.VAF.BBE'\u000d\u000aInvalid CustomField Id ='CustomField\/86', FieldName = 'CX.VAF.ABE'\u000d\u000aInvalid CustomField Id ='CustomField\/89', FieldName = 'CX.VAF.BAY'\u000d\u000aInvalid CustomField Id ='CustomField\/99', FieldName = 'CX.VAF.AAY'\u000d\u000aInvalid CustomField Id ='CustomField\/103', FieldName = 'CX.VAF.AAS'\u000d\u000aInvalid CustomField Id ='CustomField\/121', FieldName = 'CX.BORROWER.PAIRS'\u000d\u000aCustomField = [CX.ALL.FACTORS.EVAL] value = '' cannot be converted to type ='X'\u000d\u000aCustomField = [CX.DOC.ASSETSEXPIRE] value = '' cannot be converted to type ='DATE'\u000d\u000aCustomField = [CX.DOC.CREDITEXPIRES] value = '' cannot be converted to type ='DATE'\u000d\u000aCustomField = [CX.DOC.INCOMEEXPIRES] value = '' cannot be converted to type ='DATE'\u000d\u000aCustomField = [CX.DOC.OTHEREXPIRES] value = '' cannot be converted to type ='DATE'\u000d\u000aCustomField = [CX.DOC.EXPIRES] value = '' cannot be converted to type ='DATE'\u000d\u000aCustomField = [CX.UW.DECLINE.DATE] value = '' cannot be converted to type ='DATE'\u000d\u000aCustomField = [CX.NOI.SENT.DT] value = '' cannot be converted to type ='DATE'\u000d\u000aCustomField = [CX.SMS.INV.DELIVERY.DT] value = '' cannot be converted to type ='DATE'\u000d\u000aCustomField = [CX.SMS.INV.SRP] value = '' cannot be converted to type ='DECIMAL_3'\u000d\u000aCustomField = [CX.MCC.RATE] value = '' cannot be converted to type ='DECIMAL_3'\u000d\u000aCustomField = [CX.COUPON.RATE] value = '' cannot be converted to type ='DECIMAL_3'\u000d\u000aCustomField = [CX.PURCH.ADVICE.PERDIEM.DAYS] value = '' cannot be converted to type ='INTEGER'\u000d\u000aCustomField = [CX.PURCH.ADVICE.PERDIEM.RATE] value = '' cannot be converted to type ='DECIMAL_3'\u000d\u000aCustomField = [CX.CC.BORROWER.COUNT] value = '' cannot be converted to type ='INTEGER'\u000d\u000aCustomField = [CX.DOC.APPRAISAL.EFFDATE] value = '' cannot be converted to type ='DATE'\u000d\u000aCustomField = [CX.RESET] value = '' cannot be converted to type ='DATE'\u000d\u000aCustomField = [CX.RESET.OVERRIDE] value = '' cannot be converted to type ='YN'\u000d\u000a"}

**Comment by TylerBrinkley** [comment]
Hmm, that's interesting. That appears to be a different issue. I'm working on adding dirty checking support for collections. I'll let you know how that goes.

**Comment by walkeryan** [comment]
Ok great, I'll move on to another section and check back later. Thank you.

**Comment by hfcjweinstock** [comment]
For the custom fields, you may want to update the ID and name to force that to be part of the collection.

**Comment by walkeryan** [comment]
When you say the update ID and Name do you mean like:
cf.FieldName = cf.FieldName;
cf.Id = cf.Id;

**Comment by TylerBrinkley** [comment]
Also you can new up a `Loan` to serialize just what you want to send.

```c#
var loan = new Loan(client, loanId)
{
    CustomFields = new List<CustomField>
    {
        new CustomField { FieldName = "CUST91FV", StringValue = "Delete this!" }
    }
}
```

Note: this will probably break with the updates to support collection dirty-checking as collections will be a custom List type.

**Comment by hfcjweinstock** [comment]
ryan : yes.  the custom fields are update base on 1 of those 2 fields.

**Comment by TylerBrinkley** [comment]
Hmm, sounds like we need to put some logic in `CustomField` to only serialize `FieldName` if both properties have values.

**Comment by walkeryan** [comment]
I tried using the the FieldName and ID assign to itself but sadly its not setting that to update. I've tried the new loan approach but I get a null reference error when trying to assign it.

Loan loan = new Loan(client, item.LoanGuid);
                            CustomField cf = new CustomField
                            {
                                FieldName = "CUST91FV",
                                StringValue = "Delete this bad boy!"
                            };
                            loan.CustomFields.Add(cf);

Results in: -		InnerException	{"Object reference not set to an instance of an object."}	System.Exception {System.NullReferenceException}

**Comment by TylerBrinkley** [comment]
That's because CustomFields is null. I'm going to be updating collections and entity references to new up references in case they're null upon getting the property.

**Comment by walkeryan** [comment]
Success, I was able to update the Custom Field. Once you figure out the Collection Serialization I'll update my code but this will get me going for now. Thank you both for the help.

Here is what I used:

Loan loan = new Loan(client, item.LoanGuid);

var CustomFields = new List<CustomField>
 {
       new CustomField { FieldName = "CUST91FV", StringValue = "Delete this!" }
};
loan.CustomFields = CustomFields;
await client.Loans.UpdateLoanAsync(loan, true).ConfigureAwait(false);
